projects/igniteui-angular/src/lib/data-operations/data-util.ts

```diff
@@ -181,60 +180,122 @@ export class DataUtil {
-            }
             if (transaction && transaction.type === TransactionType.UPDATE) {
                 data[index] = transaction.newValue;
             }
         });
 
+        if (deleteRows) {
+            transactions
+                .filter(t => t.type === TransactionType.DELETE)
+                .forEach(t => {
+                    const index = primaryKey ? data.findIndex(d => d[primaryKey] === t.id) : data.findIndex(d => d === t.id);
+                    if (0 <= index && index < data.length) {
+                        data.splice(index, 1);
+                    }
+                });
+        }
+
         data.push(...transactions
             .filter(t => t.type === TransactionType.ADD)
             .map(t => t.newValue));
+
         return data;
     }
 
-    // TODO: optimize addition of added rows. Should not filter transaction in each recursion!!!
-    /** @experimental @hidden */
+    /**
+     * Merges all changes from provided transactions into provided hierarchical data collection
+     * @param data Collection to merge
+     * @param transactions Transactions to merge into data
+     * @param childDataKey Data key of child collections
+     * @param primaryKey Primary key of the collection, if any
+     * @param deleteRows Should delete rows with DELETE transaction type from data
+     * @returns Provided data collections updated with all provided transactions
+     */
     public static mergeHierarchicalTransactions(
         data: any[],
         transactions: HierarchicalTransaction[],
         childDataKey: any,
         primaryKey?: any,
-        parentKey?: any): any[] {
-
-        for (let index = 0; index < data.length; index++) {
-            const dataItem = data[index];
-            const rowId = primaryKey ? dataItem[primaryKey] : dataItem;
-            const updateTransaction = transactions.filter(t => t.type === TransactionType.UPDATE).find(t => t.id === rowId);
-            const addedTransactions = transactions.filter(t => t.type === TransactionType.ADD).filter(t => t.parentId === rowId);
-            if (updateTransaction || addedTransactions.length > 0) {
-                data[index] = mergeObjects(cloneValue(dataItem), updateTransaction && updateTransaction.newValue);
-            }
-            if (addedTransactions.length > 0) {
-                if (!data[index][childDataKey]) {
-                    data[index][childDataKey] = [];
+        deleteRows: boolean = false): any[] {
+
+        for (let i = 0; i < transactions.length; i++) {
+            const transaction = transactions[i];
+            const path = transaction.path;
+
+            if (path) {
+                //  We need to get parent data row. If there is a path and path contains this row id,
+                //  this is the case for UPDATE and DELETE transactions type, remove the last id from
+                //  the path
+                if (path.find(id => id === transaction.id)) {
+                    path.splice(-1, 1);
```

Comment:
Can't really find a case when this is executed, but wouldn't it edit the actual transaction?
Suggesting the transaction path wasn't correct in the first place maybe?

Reply:
yes, fixed!

Reply:
actually with last changes we do not need that check - it is removed now!